File: py_solutions/27.py
def appplyMask(mask, number):
    mask1 = int(mask.replace("X", "0"), 2)
    mask0 = int(mask.replace("X", "1"), 2)
    
    return (number | mask1) & mask0

filepath = 'input_14.txt'
addressDict = {}
with open(filepath) as fp:
   line = fp.readline()
   cnt = 1
   mask = ""
   while line:
       if "mask" in line:
           print("Setting mask to " + line[7:])
           mask = line[7:]
       else:
            splitLines = line.split(" = ")
            address = splitLines[0][4:-1]
            number = int(splitLines[1])

            masked = appplyMask(mask, number)

            addressDict[address] = masked
            print(" number " + number.__str__() + ", address " + address + ", masked " + masked.__str__())
       line = fp.readline()
       cnt += 1

   sum = 0
   for k, v in addressDict.items():
    sum += v
   
   print("sum = " + sum.__str__())

File: js_solutions/27.js
const fs = require('fs');

String.prototype.replaceAt = function(index, replacement) {
  return this.substr(0, index) + replacement + this.substr(index + replacement.length);
}

function applyMask(number, mask) {
  const mask1 = BigInt(`0b${mask.replace(/X/g, "0")}`);
  const mask0 = BigInt(`0b${mask.replace(/X/g, "1")}`);

  return (number | mask1) & mask0;
}

fs.readFile('input_14.txt', 'utf8' , (err, data) => {
  if (err) {
    console.error(err)
    return
  }

  var lines = data.trim().split("\n");

  var mask;
  var numericArray = {};

  for(let i = 0; i < lines.length; i++) {
    if(lines[i].indexOf("mask") > -1) {
      mask = lines[0].split("mask = ")[1]
    } else {
      const setParts = lines[i].split(" = ");
      const address = parseInt(setParts[0].slice(4).slice(0, -1));
      const number = BigInt(setParts[1]);

      const masked = applyMask(number, mask);
      numericArray[address] = masked;

      console.log(`number: ${number}, address = ${address}, masked: ${masked}`);
    }
  }

  var sum = 0n;
  for (const [key, value] of Object.entries(numericArray)) {
    sum += value;
  }

  console.log(sum);
})
